### ISSUE TITLE
edgeAgent apply CoolOff after closed-but-regular reboots

### ISSUE THREAD

**Issue by FabGre** [Azure/iotedge#3702]
Hi there. We are working on an iotedge device which has a specific "low-power" behaviour : it has to be off most of the time and start-up only a few times per days and for a few minutes to collect all its sensors (with different custom edge modules), sent data to iotHub, then shutdown.

During our dev, we start-up again and again our device, always for very few minutes, and we saw that the edgeAgent module applies increasing CoolOff to all modules, including edgeHub module when the reboots are too close. We perform regular system shutdown so systemd should nicely close iotedge stack.

**After "a while", edgeAgent will not start any module anymore**, which is a bug issue for us.

## Expected Behavior
We think that a regular shutdown should not be considered as an anormal situation or, at least, we should have some configuration parameter to define what we consider normal.

## Current Behavior
edgeAgent applies CoolOff to all modules if start-up/shutdown are too close.

## Steps to Reproduce
It's not so hard to reproduce by creating an infinite reboot loop : 
- you need an ubuntu Virtual Machine with iotedge 1.0.9 installed and device enrolled in an Azure IoTHub
- apply a default empty module manifest like the one included in this Issue (our manifest without custom modules) : 
[empty-deployment.json.txt](https://github.com/Azure/iotedge/files/5365378/empty-deployment.json.txt)
- OPTIONAL use this docker daemon config to send all docker logs to syslog (easy to follow a start-up cycle) : 

```json
{
  "log-driver": "syslog",
  "log-opts": {
    "syslog-address": "udp://localhost:514"
  }
}
```

and allow rsyslog to received them by uncomment this in `/etc/rsyslog.conf`

```
# provides UDP syslog reception
module(load="imudp")
input(type="imudp" port="514")
```

- create a reboot script like this one : 

```bash
#!/bin/sh

shutdown -r +2 "Test cycle reboots"
```

Warn : with this way, you will not be able to connect to your VM after shutdown is performed. A sleep before shutdown command could be convenient.

- execute at each start-up the script, for example with a ExecStartPost on iotedge systemd service `/lib/systemd/system/iotedge.service`  : 

```
[Service]
ExecStart=/usr/bin/iotedged -c /etc/iotedge/config.yaml
ExecStartPost=+/home/USER/test_cooloff_reboot.sh
KillMode=process
TimeoutStartSec=600
...
```

## Context (Environment)

We experienced this behaviour with both Ubuntu x86 Virtual Machine and our custom Yocto (Thud) ARM device. We are using last 1.0.9 iotedge.

### Runtime Versions
* iotedged [1.0.9.5]: 
* Edge Agent [1.0.9]: 
* Edge Hub [1.0.9]: 
* Docker/Moby [run `docker version`]: 

```
Client:
 Version:           19.03.13+azure
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        cd8016b6bcc51a51f577bcffb37e6a870f7813f9
 Built:             Tue Sep 15 23:02:04 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.13+azure
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       bd33bbf0497b2327516dc799a5e541b720822a4c
  Built:            Mon Mar 12 00:00:00 2018
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.7+azure
  GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.18.0
  GitCommit:
```
Note: when using Windows containers on Windows, run `docker -H npipe:////./pipe/iotedge_moby_engine version` instead

## Logs
<details>
<summary>edge-agent logs</summary>

```
Oct 12 13:38:32 localhost edgeAgent[759]: <6> 2020-10-12 13:38:32.373 +00:00 [INF] - Module 'edgeHub' scheduled to restart after 05m:00s (55s left).
```
</details>

## Additional Information
Let me kwon if you need more/specific logs.

**Comment by massand** [comment]
> After "a while", edgeAgent will not start any module anymore, which is a bug issue for us.

@FabGre could you confirm if edgeAgent applies a CoolOff > 5 minutes before starting any module? Or is it capped at 5 minutes and that is above your device's cutoff threshold?

**Comment by FabGre** [comment]
@massand Hello, yes your right we have a system watchdog which shutdown before the end of the maximum 5min CoolOff value. **But**, "after a while", we definitely saw that there's no module start-up triggered by the edgeModule in our syslogs.

We tried several things when it's happened :

**Comment by massand** [comment]
I assume your system watchdog would need to allow edgeAgent between 5-10 minutes (maximum cool off period for edgeAgent's exponential backoff retry mechanism for issuing docker management commands) between restarts. This would be a feature request if less time is provided to edgeAgent in one boot session.

If enough time is being provided to edgeAgent and this is still an issue, we need more information because it isn't clear what is causing the `..Module 'edgeHub' scheduled to restart..` logs. Is it really a `docker pull` command that is being retried?

Please share the edgeAgent (set RuntimeLogLevel=debug) and iotedged logs. Here is a bash command to enable debug logging for iotedged:
```
bash -c "echo '[Service]
Environment=IOTEDGE_LOG=edgelet=debug' > /etc/systemd/system/iotedge.service.d/override.conf"
```
Please provide a set of logs just before each system watchdog shutdown (where edgeAgent is expected to retry failed commands few times with exponential backoff). Also, please attach a second set of edgeAgent and iotedged logs when edgeAgent stops triggering module commands subsequently after several retries.

**Comment by FabGre** [comment]
Hi @massand, sorry I can't find time to perform the tests you asked but two questions remains :

* why do the edgeAgent apply cool off to modules after a clean shutdown ?
* why can it falls in a mode when it never starts module anymore ?

You can easily test that with a VM, instructions and script I provided in my first post. Thx

**Comment by massand** [comment]
> why do the edgeAgent apply cool off to modules after a clean shutdown ? (which cause the problem in the first place)

As per edgeAgent's [RestartPolicyManager](https://github.com/Azure/iotedge/blob/master/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/RestartPolicyManager.cs#L76),  `max(2 ^ restartCount, maxCoolOff)` is applied to each module regardless of when it is started. 

> why can it falls in a mode when it never starts module anymore ?

Without module logs that are failing, it is hard for me to tell what the issue is.

**Comment by massand** [comment]
@veyalla provided me the possible RCA and workaround for edgeAgent never starting modules anymore - 
>I believe this is controlled by MaxRestartCount env var which is set to a default value of 20. As a temporary workaround, this value can be set to a very large number. This was documented recently in our repo: https://github.com/Azure/iotedge/blob/master/doc/EnvironmentVariables.md

**Comment by pcurt** [comment]
I am working with @FabGre
Looking  at sources, I have applied the following configuration:
-  `IntensiveCareTimeInMinutes = 0` To always clear the module stats
- `CoolOffTimeUnitInSeconds = 1` To only wait 2 second after a backOff module state
- `MaxRestartCount = 2147483647`  To never block a module start

We are now using the following versions :
- iotedged [1.0.10]
- Edge Agent [1.0.10]
- Edge Hub [1.0.10]

I am now checking that there is no side effect with this configuration

**Comment by FabGre** [comment]
Hello Venkat, for now we are testing values for parameters found by @pcurt with the 1.0.10 version which is now available for our Yocto Thud (_Thx Azure team for that !_). Seems it works as expected.

**Comment by varunpuranik** [comment]
Cool, thanks for the update @FabGre. Glad to hear that it is working fine for you now. I will close this issue. Feel free to open a new issue if needed.
